docs/design/persistent-volume-provisioning.md

```diff
@@ -0,0 +1,129 @@
+	BoundReplicas int `json:"replicas"`
+	AvailableReplicas int `json:"replicas"`
+}
+
+// PersistentVolumeTemplateSpec describes the persistent volume to be created from a template
+type PersistentVolumeTemplateSpec struct {
+	// Metadata of the PVs created from this template.
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines a persistent volume.
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+}
+
+```
+
+##### PersistentVolumeController API
+
+| Action | HTTP Verb | Path | Description |
+| ---- | ---- | ---- | ---- |
+| CREATE | POST | /api/{version}/persistentvolumecontrollers/ | Create instance of PersistentVolumeController|
+| GET | GET | /api/{version}persistentvolumecontrollers/{name} | Get instance of PersistentVolumeController |
+| UPDATE | PUT | /api/{version}/persistentvolumecontrollers/{name} | Update instance of PersistentVolumeController |
+| DELETE | DELETE | /api/{version}/persistentvolumecontrollers/{name} | Delete instance of PersistentVolumeController |
+| LIST | GET | /api/{version}/persistentvolumecontrollers | List instances of PersistentVolumeController|
+| WATCH | GET | /api/{version}/watch/persistentvolumecontrollers | Watch for changes to a PersistentVolumeController|
+
+
+#### PersistentVolumeControllerManager
+
+Use DeltaFIFO to watch PersistentVolumes of a particular type as defined by the controller's label selector.
+
+Add/Sync events inspect for binding (pv.Spec.ClaimRef) and tallies available/bound.
+
+Uupdate events inspect for binding and phase.  Any volume unbound from its claim will be available for recycling.
+
+```
+if controller.replicas > 0:
+    while availableCount < controller.replicas:
+        make(controller.spec.template)
+
+    while availableCount > controller.replicas:
+        recycle(v) // where v is choosen at random
+        
+for v in listOfVolumesToRecycle: //pv.Status.Phase = api.VolumeReleased
+        recycle(v)
+```
+
+NFS is a good example where controller.replicas will equal 0.  Dynamic creation will not take place, but the controller
+is still available to recycle volumes.  The controller is required to manage the complete lifecycle of the NFS volume.
+
+#### Volumes and their plugins
+
+Use of cloud and non-cloud volumes as examples for controllers.
+
+| Volume |  Dynamic Creation | Formatting | Reclamation |
+| ---- | ---- | ---- | ---- |
+| GCE | yes via API | mount volume, uses "safeFormatAndMount" func | delete via API |
+| AWS | yes via API | mount volume, uses "safeFormatAndMount" func | delete via API |
+| OpenStack | yes via API | mount volume, uses "safeFormatAndMount" func | delete via API |
+| NFS | No, externally provisioned | pre-formatted during provisioning | mount volume, run "rm -rf" on volume |
+| ISCSI | ? | ? | ? |
+| gluster | ? | ? | ? |
+| ceph | ? | ? | ? |
+
+
+#### Limitations
+
+The controller should not be mounting volumes, formatting disks, or other tasks associated with provisioning.  Ideally, a job
```

Comment:
You can either require the admin to run a pod and then you communicate with that pod or you can run pods yourself.  Running pods will not be able to format a disk, but they should be able to rm on it.

Our support for pods that complete is overall weak right now